### PR TITLE
Add weighted_mean

### LINKED ISSUE #937: Add weighted_mean function
I think it will be helpful if we add an option `:sample_weights`. It will default to `nil` and then it will calculate the arithmetic mean. Otherwise, it will calculate the weighted mean. We can add the analogical functionality to `Nx.mean`.

### PR DESCRIPTION
**Pull request by msluszniak**
Closes #937

I implemented a version which accepts only one axis to aggregate. That's because it is unclear what this function would do for more than one dimension.

### CODE REVIEW

**Comment by polvalente** [comment]
https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.average.html
Jax (as well as NumPy) seem to only accept a single axis, so no worries about that!

File: nx/lib/nx.ex [hunk 1]
```diff
@@ -6612,6 +6612,152 @@ defmodule Nx do
   defp mean_den(shape, [axis | axes]) when axis >= 0,
     do: elem(shape, axis) * mean_den(shape, axes)
 
+  @doc """
+  Returns the weighted mean for the tensor and the weights.
+
+  If the `:axis` option is given, it aggregates over
+  that dimension, effectively removing it. `axes: [0]`
```
Comment (polvalente) [inline nx/lib/nx.ex:6619]:
```suggestion
  that dimension, effectively removing it. `axis: 0`
```

File: nx/lib/nx.ex [hunk 2]
```diff
@@ -6612,6 +6612,152 @@ defmodule Nx do
...
+  implies aggregating over the highest order dimension
+  and so forth. If the axis is negative, then counts
+  the axis from the back. For example, `axis: [-1]` will
```
Comment (polvalente) [inline nx/lib/nx.ex:6622]:
```suggestion
  and so forth. If the axis is negative, then the axis will
  be counted from the back. For example, `axis: -1` will
```

File: nx/lib/nx.ex [hunk 3]
```diff
@@ -6612,6 +6612,152 @@ defmodule Nx do
...
+  always aggregate all rows. You can only aggregate over
+  one axis.
```
Comment (polvalente) [inline nx/lib/nx.ex:6624]:
```suggestion
  always aggregate over the last dimension.
```

We shouldn't use rows for N-dim tensors. Also, "axis" being singular is our standard for "single-axis aggregation"

File: nx/lib/nx.ex [hunk 4]
```diff
@@ -6612,6 +6612,152 @@ defmodule Nx do
...
+
+  You may optionally set `:keep_axis` to true, which will
```
Comment (polvalente) [inline nx/lib/nx.ex:6626]: I'm not sure if we should use keep_axis or keep_axes here. Both would make sense in that keep_axis -> "keep the aggregated axis" and keep_axes -> "keep all axes present in the original tensor"

File: nx/lib/nx.ex [hunk 5]
```diff
@@ -6612,6 +6612,152 @@ defmodule Nx do
...
+        ]
+      >
+
+      iex> t = Nx.tensor([[[1, 2, 3], [4, 5, 6]], [[7, 8, 9], [10, 11, 12]]], names: [:x, :y, :z])
+      iex> weights = Nx.tensor([[[8, 2, 4], [0, 3, 5]], [[4, 2, 6], [3, 1, 5]]])
+      iex> Nx.weighted_mean(t, weights, axis: -1)
+      #Nx.Tensor<
+        f32[x: 2][y: 2]
+        [
+          [1.7142857313156128, 5.625],
```
Comment (polvalente) [inline nx/lib/nx.ex:6675]: I think we should have cleaner numbers in the example.
I suggest having a weights tensor with lots of 1s, -1s and 0s so that it's easier to calculate everything

File: nx/lib/nx.ex [hunk 6]
```diff
@@ -6612,6 +6612,152 @@ defmodule Nx do
...
+        else: opts[:axis]
+
+    weights =
+      if shape != weights_shape do
+        cond do
+          axis == nil ->
+            raise ArgumentError, "Axis must be specified when shapes of input and weights differ."
+
+          tuple_size(weights_shape) != 1 ->
+            raise ArgumentError, "1D weights expected when shapes of a and weights differ."
```
Comment (polvalente) [inline nx/lib/nx.ex:6717]:
```suggestion
            raise ArgumentError, "1D weights expected when input shapes differ."
```

File: nx/lib/nx.ex [hunk 7]
```diff
@@ -6612,6 +6612,152 @@ defmodule Nx do
...
+    axis =
+      if opts[:axis] != nil,
+        do: Nx.Shape.normalize_axes(shape, [opts[:axis]], names),
+        else: opts[:axis]
+
+    weights =
+      if shape != weights_shape do
+        cond do
+          axis == nil ->
+            raise ArgumentError, "Axis must be specified when shapes of input and weights differ."
```
Comment (polvalente) [inline nx/lib/nx.ex:6714]:
```suggestion
            raise ArgumentError, "axis must be specified when shapes of input and weights differ."
```

File: nx/lib/nx.ex [hunk 8]
```diff
@@ -6612,6 +6612,152 @@ defmodule Nx do
...
+
+          tuple_size(weights_shape) != 1 ->
+            raise ArgumentError, "1D weights expected when shapes of a and weights differ."
+
+          elem(weights_shape, 0) != elem(shape, axis) ->
+            raise ArgumentError, "Length of weights not compatible with specified axis."
```
Comment (polvalente) [inline nx/lib/nx.ex:6720]:
```suggestion
            raise ArgumentError, "length of weights not compatible with specified axis."
```

File: nx/lib/nx.ex [hunk 9]
```diff
@@ -6612,6 +6612,152 @@ defmodule Nx do
...
+
+          true ->
+            nil
+        end
+
+        dims_to_broadcast =
+          for(_ <- 0..(tuple_size(shape) - 2), do: 1) ++ Tuple.to_list(weights_shape)
```
Comment (polvalente) [inline nx/lib/nx.ex:6727]:
```suggestion
          List.duplicate(1, tuple_size(shape) - 1) ++ Tuple.to_list(weights_shape)
```

File: nx/lib/nx.ex [hunk 10]
```diff
@@ -6612,6 +6612,152 @@ defmodule Nx do
...
+
+        dims_to_broadcast = List.to_tuple(dims_to_broadcast)
+        weights = broadcast(weights, dims_to_broadcast)
+        dims_to_swap = for i <- 0..(tuple_size(shape) + tuple_size(weights_shape) - 2), do: i
```
Comment (polvalente) [inline nx/lib/nx.ex:6731]:
```suggestion
        dims_to_swap = Enum.to_list(0..(tuple_size(shape) + tuple_size(weights_shape) - 2))
```

File: nx/lib/nx.ex [hunk 11]
```diff
@@ -6612,6 +6612,152 @@ defmodule Nx do
...
+
+        dims_to_swap = swap(dims_to_swap, axis, tuple_size(shape) + tuple_size(weights_shape) - 2)
+
+        transpose(weights, axes: dims_to_swap)
```
Comment (polvalente) [inline nx/lib/nx.ex:6735]: why is this transpose needed?

File: nx/lib/nx.ex [hunk 12]
```diff
@@ -6612,6 +6612,152 @@ defmodule Nx do
...
+      else
+        weights
+      end
+
+    weights_sum = sum(weights, axes: axis, keep_axes: opts[:keep_axis])
+
+    if to_number(any(equal(weights_sum, 0.0))) == 1 do
+      raise ArgumentError, "Weights sum to zero, can't be normalized"
+    end
```
Comment (polvalente) [inline nx/lib/nx.ex:6744]: I would just let it crash in the operation. This check will take the operation from the device instead of running everything in-device

File: nx/lib/nx.ex [hunk 13]
```diff
@@ -6612,6 +6612,152 @@ defmodule Nx do
...
+
+    divide(
+      sum(multiply(tensor, weights), axes: axis, keep_axes: opts[:keep_axis]),
+      weights_sum
+    )
```
Comment (polvalente) [inline nx/lib/nx.ex:6749]: Please turn this into a pipeline

File: nx/lib/nx.ex [hunk 14]
```diff
@@ -6612,6 +6612,152 @@ defmodule Nx do
...
+  @doc type: :aggregation, from_backend: false
+  def weighted_mean(tensor, weights, opts \\ []) do
+    opts = keyword!(opts, axis: nil, keep_axis: false)
+    %T{shape: shape, names: names} = tensor = to_tensor(tensor)
+    %T{shape: weights_shape} = weights = to_tensor(weights)
+
+    axis =
+      if opts[:axis] != nil,
+        do: Nx.Shape.normalize_axes(shape, [opts[:axis]], names),
+        else: opts[:axis]
```
Comment (polvalente) [inline nx/lib/nx.ex:6708]: either call the variable `axes` or extract it into a number | nil again. I found it confusing that the singular axis refers to a list here

File: nx/lib/nx.ex (hunk 4 above)
Comment (msluszniak) [reply nx/lib/nx.ex:6625]: I agree. What do you think @josevalim?

Comment (josevalim) [reply nx/lib/nx.ex:6625]: We have both `keep_axis` and `keep_axes` in Nx, so I think it should be `keep_axis` when the other option is `axis` (and `keep_axes` for `axes`).

File: nx/lib/nx.ex [hunk 15]
```diff
@@ -6612,6 +6612,141 @@ defmodule Nx do
...
+      if shape != weights_shape do
+        cond do
+          axes == nil ->
+            raise ArgumentError, "axis must be specified when shapes of input and weights differ."
+
+          tuple_size(weights_shape) != 1 ->
+            raise ArgumentError, "rank-1 weights tensor expected when input shapes differ."
+
+          elem(weights_shape, 0) != elem(shape, List.first(axes)) ->
+            raise ArgumentError, "length of weights not compatible with specified axis."
```
Comment (josevalim) [inline nx/lib/nx.ex:6727]:
```suggestion
            raise ArgumentError, "axis must be specified when shapes of input and weights differ"

          tuple_size(weights_shape) != 1 ->
            raise ArgumentError, "rank-1 weights tensor expected when input shapes differ"

          elem(weights_shape, 0) != elem(shape, List.first(axes)) ->
            raise ArgumentError, "length of weights not compatible with specified axis"
```

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:
